### PR TITLE
GC: update required verbs for deletable resources, allow list of ignored resources to be customized

### PR DESCRIPTION
**Pull request by ncdc**
The garbage collector controller currently needs to list, watch, get,
patch, update, and delete resources. Update the criteria for
deletable resources to reflect this.

Also allow the list of resources the garbage collector controller should
ignore to be customizable, so downstream integrators can add their own
resources to the list, if necessary.

cc @caesarxuchao @deads2k @smarterclayton @mfojtik @liggitt @sttts @kubernetes/sig-api-machinery-pr-reviews

### CODE REVIEW

**Comment by mfojtik** [comment]
@ncdc nicely done! code lgtm.

File: pkg/controller/garbagecollector/graph_builder.go [hunk 1]
```diff
@@ -243,6 +244,13 @@ var ignoredResources = map[schema.GroupVersionResource]struct{}{
 	{Group: "apiregistration.k8s.io", Version: "v1alpha1", Resource: "apiservices"}:            {},
 }
 
+// IgnoredResources returns the default set of resources that the garbage collector controller
+// should ignore. This is exposed so downstream integrators can have access to the defaults, and add
+// to them as necessary when constructing the controller.
+func IgnoredResources() map[schema.GroupVersionResource]struct{} {
```
Comment (caesarxuchao) [inline pkg/controller/garbagecollector/graph_builder.go]: Rename to DefaultIgnoredResources()?

**Comment by caesarxuchao** [comment]
Last two commits lgtm except for a nit.

File: pkg/controller/garbagecollector/graph_builder.go (hunk 1 above)
Comment (ncdc) [inline pkg/controller/garbagecollector/graph_builder.go]: Updated

**Comment by soltysh** [comment]
lgtm for the last two commits.

**Comment by ncdc** [comment]
Rebased on top of master now that #45427 is in. @deads2k tag if this looks good to you?

File: cmd/kube-controller-manager/app/core.go [hunk 2]
```diff
@@ -178,7 +178,7 @@ func startGarbageCollectorController(ctx ControllerContext) (bool, error) {
 	if err != nil {
 		return true, fmt.Errorf("failed to get supported resources from server: %v", err)
 	}
-	deletableResources := discovery.FilteredBy(discovery.SupportsAllVerbs{Verbs: []string{"delete"}}, preferredResources)
+	deletableResources := discovery.FilteredBy(discovery.SupportsAllVerbs{Verbs: []string{"list", "watch", "delete"}}, preferredResources)
```
Comment (deads2k) [inline cmd/kube-controller-manager/app/core.go]: Pretty sure that we use a live "get" to confirm existence and "patch" to orphan resources.

Comment (ncdc) [inline cmd/kube-controller-manager/app/core.go]: So you want to make sure that for any resource we might want to gc or orphan, that we can get/list/watch/patch/delete?

File: cmd/kube-controller-manager/app/core.go [hunk 3]
```diff
@@ -189,7 +189,14 @@ func startGarbageCollectorController(ctx ControllerContext) (bool, error) {
 	metaOnlyClientPool := dynamic.NewClientPool(config, restMapper, dynamic.LegacyAPIPathResolverFunc)
 	config.ContentConfig = dynamic.ContentConfig()
 	clientPool := dynamic.NewClientPool(config, restMapper, dynamic.LegacyAPIPathResolverFunc)
-	garbageCollector, err := garbagecollector.NewGarbageCollector(metaOnlyClientPool, clientPool, restMapper, deletableGroupVersionResources, ctx.InformerFactory)
+	garbageCollector, err := garbagecollector.NewGarbageCollector(
+		metaOnlyClientPool,
+		clientPool,
+		restMapper,
+		deletableGroupVersionResources,
+		garbagecollector.DefaultIgnoredResources(),
```
Comment (deads2k) [inline cmd/kube-controller-manager/app/core.go]: I think this should be option on the config struct contained in the controller context.  It's logically config information even if its never exposed via flags.  You can set the default where its created.

File: cmd/kube-controller-manager/app/core.go (hunk 2 above)
Comment (deads2k) [inline cmd/kube-controller-manager/app/core.go]: > So you want to make sure that for any resource we might want to gc or orphan, that we can get/list/watch/patch/delete?

Yes

File: cmd/kube-controller-manager/app/core.go (hunk 3 above)
Comment (ncdc) [inline cmd/kube-controller-manager/app/core.go]: @deads2k like this? https://gist.github.com/ncdc/cc366247ebda5b434a3da0668637613d

File: cmd/kube-controller-manager/app/core.go (hunk 2 above)
Comment (ncdc) [inline cmd/kube-controller-manager/app/core.go]: I just looked and we also use "update" to remove the finalizer (there is a todo to switch to patch when strategic merge supports removing an entry from a slice of a base type). I'll make the appropriate changes.

**Comment by ncdc** [comment]
@deads2k updated

**Comment by deads2k** [comment]
You'll need to define a serializeable GroupResource, but yes.

**Comment by ncdc** [comment]
> You'll need to define a serializeable GroupResource, but yes.

What do you mean / when / where?

**Comment by deads2k** [comment]
> What do you mean / when / where?

Down inside KubeControllerManagerConfiguration

**Comment by ncdc** [comment]
I didn't make it part of `KubeControllerManagerConfiguration`. I made it part of the CMServer struct.

**Comment by ncdc** [comment]
@deads2k if it needs to be in `KubeControllerManagerConfiguration`, how do you propose making a GVR serializable?

**Comment by ncdc** [comment]
Updated, now with GCIgnoredResources living in the `KubeControllerManagerConfiguration` componentconfig struct.

**Comment by deads2k** [comment]
lgtm

please squash before merge.

**Comment by ncdc** [comment]
Squashed

**Comment by caesarxuchao** [comment]
lgtm

**Comment by ncdc** [comment]
@k8s-bot pull-kubernetes-kubemark-e2e-gce test this

**Comment by ncdc** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this (took too long to delete some namespaces).

**Comment by smarterclayton** [comment]
/approve

based on reviewers signing off

**Pull request merged**